File: codes/backtraking/78.subsets/subsets.go
package main

func subsets(nums []int) [][]int {
	var result [][]int
	var path []int
	var backtracing func(int)

	backtracing = func(beg int) {
		if beg == len(nums) {
			return
		}
		for i := beg; i < len(nums); i++ {
			path = append(path, nums[i])
			result = append(result, append([]int{}, path...))
			backtracing(i + 1)
			path = path[:len(path)-1]
		}
	}

	result = append(result, []int{})
	backtracing(0)
	return result
}

File: codes/backtraking/78.subsets/subsets.cpp
#include <vector>
using namespace std;
class Solution {
    using vit = vector<int>::const_iterator;
    vector<vector<int>> result;
    vector<int> path;

public:
    vector<vector<int>> subsets(vector<int> &nums) {
        result.push_back(vector<int>{});
        backtracing(nums.cbegin(), nums.cend());
        return result;
    }

    void backtracing(vit beg, vit end) {
        if (beg == end) {
            return;
        }
        for (auto it = beg; it != end; ++it) {
            path.push_back(*it);
            result.push_back(path);
            backtracing(it + 1, end);
            path.pop_back();
        }
    }
};